crates/sui-types/src/attestation.rs

```diff
@@ -0,0 +1,539 @@
+                    "Cannot parse PCRs".to_string(),
+                ))
+            }
+        };
+
+        let user_data = get_optional_bytes("user_data");
+        let nonce = get_optional_bytes("nonce");
+
+        let cabundle = match document_map
+            .iter()
+            .find(|(k, _)| matches!(k, ciborium::value::Value::Text(s) if s == "cabundle"))
+        {
+            Some((_, ciborium::value::Value::Array(arr))) => arr
+                .iter()
+                .map(|v| match v {
+                    ciborium::value::Value::Bytes(bytes) => Ok(bytes.clone()),
+                    _ => Err(NitroError::InvalidAttestationDoc(
+                        "Invalid cabundle format".to_string(),
+                    )),
+                })
+                .collect::<Result<Vec<_>, _>>()?,
+            _ => {
+                return Err(NitroError::InvalidAttestationDoc(
+                    "Cannot parse cabundle".to_string(),
+                ))
+            }
+        };
+
+        Ok(AttestationDocument {
+            module_id,
+            timestamp,
+            digest,
+            pcrs,
+            certificate,
+            cabundle,
+            public_key,
+            user_data,
+            nonce,
+        })
+    }
+
+    /// Verify the certificate against AWS Nitro root of trust and checks expiry.
+    fn validate_cert(&self, now: u64) -> Result<(), NitroError> {
+        let mut cert_chain: Vec<CertificateDer<'static>> = Vec::new();
+        for ca_cert in self.cabundle.iter().rev() {
+            cert_chain.push(CertificateDer::from(ca_cert.to_vec()));
+        }
+
+        let leaf_cert = CertificateDer::from(self.certificate.clone());
+        cert_chain.push(leaf_cert.clone());
+
+        // Setup root store with AWS Nitro root certificate
+        let mut root_store = RootCertStore::empty();
+        let root_cert = CertificateDer::from_pem_slice(ROOT_CERTIFICATE_PEM).map_err(|e| {
+            NitroError::InvalidCertificate(format!("Failed to parse root certificate: {}", e))
+        })?;
+        root_store
+            .add(root_cert)
+            .map_err(|e| NitroError::InvalidCertificate(e.to_string()))?;
+        let verifier = WebPkiClientVerifier::builder(root_store.into())
+            .build()
+            .map_err(|e| NitroError::InvalidCertificate(e.to_string()))?;
+
+        verifier
+            .verify_client_cert(
+                &leaf_cert,
+                &cert_chain,
+                UnixTime::since_unix_epoch(Duration::from_millis(now)),
+            )
+            .map_err(|e| NitroError::InvalidCertificate(e.to_string()))?;
+        Ok(())
+    }
+
+    /// Validate the PCRs against the expected PCRs. todo: add docs
+    fn validate_pcrs(&self, expected_pcrs: &[&[u8]]) -> Result<(), NitroError> {
+        // only pcr0, pcr1, pcr2 are checked
+        if expected_pcrs.len() != 3 {
+            return Err(NitroError::InvalidPcrs);
+        }
+        let mut is_match = true;
```

Comment:
- no need to check all pcrs if one is invalid, can simplify the code a bit
- let's check before the loop that self.pcrs.len() = 3

Reply:
the prcs returned from the enclaves can be more than 3: https://docs.aws.amazon.com/enclaves/latest/user/set-up-attestation.html#pcr012

we basically only care about equality for the first 3 since they are related to build time measurement

Reply:
why not exposing all of them from the native function and leaving it to the Move app to decide which PCRs to depend on and how?

Reply:
good point, revised.